include/zephyr/sys/util.h

```diff
@@ -201,6 +202,27 @@ extern "C" {
 		(POINTER_TO_UINT(ptr) - POINTER_TO_UINT(array)) / sizeof((array)[0]);              \
 	})
 
+/**
+ * @brief Validate if two entities have a compatible type
+ *
+ * @param a the first entity to be compared
+ * @param a the second entity to be compared
+ * @return 1 if the two elements are compatible, 0 if they are not
+ */
+#define SAME_TYPE(a, b) __builtin_types_compatible_p(__typeof__(a), __typeof__(b))
+
+/**
+ * @brief Validate CONTAINER_OF parameters, only applies to C mode.
+ */
+#ifndef __cplusplus
```

Comment:
Should this also check `HAS_BUILTIN(__builtin_types_compatible_p)`.  Do you know if __builtin_types_compatible_p is supported by all toolchains.  I know GCC and clang definitely has support.